vendor/google.golang.org/grpc/health/client.go

```diff
@@ -0,0 +1,117 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package health
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/connectivity"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/internal"
+	"google.golang.org/grpc/internal/backoff"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	backoffStrategy = backoff.DefaultExponential
+	backoffFunc     = func(ctx context.Context, retries int) bool {
+		d := backoffStrategy.Backoff(retries)
+		timer := time.NewTimer(d)
+		select {
+		case <-timer.C:
+			return true
+		case <-ctx.Done():
+			timer.Stop()
+			return false
+		}
+	}
+)
+
+func init() {
+	internal.HealthCheckFunc = clientHealthCheck
```

Comment:
I believe this actually changes the behavior of gRPC's client-side LB (which we use), such that it relies on the health check service rather than simple connectivity to judge whether a backend is up: https://pkg.go.dev/google.golang.org/grpc/internal.

I think that's fine, but note that this will only take effect for boulder components that import `google.golang.org/grpc/health`, and therefore run this init function. For instance, in this PR, boulder-ca would get the new behavior when checking its backends, but boulder-wfe would not.

I'd like to consistently do the same thing across all gRPC clients. Probably the easiest way to do that is to add an `import _ "google.golang.org/grpc/health"` to grpc/client.go, which will ensure all gRPC clients get this treatment. If you go this route, please update the PR desc to mention the changed behavior.

A slightly more conservative route would be to implement this but try to explicitly disable the client-side health checking it turns on. I believe [WithDisableHealthCheck](https://pkg.go.dev/google.golang.org/grpc#WithDisableHealthCheck) will do that, but it's marked experimental. So I think it makes sense to just take the plunge and implement checking consistently for all clients.

Reply:
Good call, done, and description updated.